### PR TITLE
fix: password reset

### PR DESCRIPTION
**Pull request by markusahlstrand**
This fixes at least the first step of the password reset flow. Not sure if it will work all the way yet.

### CODE REVIEW

File: src/routes/tsoa/login.ts [hunk 1]
```diff
@@ -447,7 +447,7 @@ export class LoginController extends Controller {
     const { env } = request.ctx;
 
     const session = await env.data.universalLoginSessions.get(state);
-    if (!session?.username) {
+    if (!session) {
```
Comment (markusahlstrand) [inline src/routes/tsoa/login.ts:450]: The user name is passed in the body, not in the session

File: src/routes/tsoa/login.ts [hunk 2]
```diff
@@ -463,23 +463,30 @@ export class LoginController extends Controller {
     }
 
     // TODO - filter by primary user
-    const [user] = await env.data.users.getByEmail(
-      client.tenant_id,
-      params.username,
-    );
+    const userResponse = await env.data.users.list(session.tenant_id, {
```
Comment (markusahlstrand) [inline src/routes/tsoa/login.ts:466]: Use the list query instead. Maybe we can remove the getByEmail? This should probably check the provider as well? And fetch the primary user if one exists?

File: src/types/Code.ts [hunk 3]
```diff
@@ -1,10 +1,9 @@
 export interface Code {
   id: string;
   code: string;
-  email: string;
```
Comment (markusahlstrand) [inline src/types/Code.ts:4]: The email isn't part of the database schema. Let's use the user_id instead.

File: src/types/Code.ts [hunk 4]
```diff
@@ -1,10 +1,9 @@
...
   type: "password_reset" | "validation";
   created_at: string;
   expires_at: string;
   used_at?: string;
-  user_id?: string;
+  user_id: string;
```
Comment (markusahlstrand) [inline src/types/Code.ts:8]: The user_Id should be mandatory

File: src/app.ts [hunk 5]
```diff
@@ -79,7 +79,7 @@ app.get("/static/:file{.*}", serve);
 app.get("/test", async (ctx: Context<{ Bindings: Env }>) => {
   const response = await ctx.env.data.applications.list(
     // This is the tenant id in dev
-    "qo0kCHUE8qAvpNPznuoRW",
+    "VpE9qtb4Gt_iCahTM0FYg",
```
Comment (markusahlstrand) [inline src/app.ts:82]: Think this might have changed. Probably this route should be removed?

Comment (danielgent) [reply src/app.ts:82]: I've never used this route! what is it? A nice exploit we've left? :smile:

File: src/routes/tsoa/login.ts (hunk 2 above)
Comment (danielgent) [reply src/routes/tsoa/login.ts:466]: https://github.com/sesamyab/auth/pull/337 I started a PR for this here :smile: 

I can go back through it and try it finish it off. Will require some discussion I think

At some point we need to be able to search by email though! Auth0 has that

File: src/routes/tsoa/login.ts [hunk 6]
```diff
@@ -463,10 +463,14 @@ export class LoginController extends Controller {
...
+      page: 0,
+      per_page: 1,
+      include_totals: false,
+      q: `email:${params.username}`,
+    });
+
+    const [user] = userResponse.users;
```
Comment (danielgent) [inline src/routes/tsoa/login.ts:473]: this is no different to what you had before though. You're still returning the first user with that email address

File: src/types/Code.ts (hunk 3 above)
Comment (danielgent) [reply src/types/Code.ts:4]: Yes we have some confusing with mixing up email & username.  :+1:

**Review by danielgent — APPROVED** [review]
I don't think we need to replace the `.getUsersByEmail` but this isn't on production so not a huge concern

File: src/routes/tsoa/login.ts (hunk 2 above)
Comment (markusahlstrand) [reply src/routes/tsoa/login.ts:466]: It is still possible to search for the email, it's just that we don't need to expose it specifically in the adapter as you can do it in the list function allready

File: src/routes/tsoa/login.ts (hunk 6 above)
Comment (markusahlstrand) [reply src/routes/tsoa/login.ts:473]: Yeah, what we want change here is the handling of linked accounts. Maybe we should have a wrapper for that? So, if you linked an account I guess you still can login using an email which isn't a primary email?

File: src/routes/tsoa/login.ts (hunk 2 above)
Comment (danielgent) [reply src/routes/tsoa/login.ts:466]: But we do it in enough places that we have the `rule-of-three` to justify creating a helper :smile: 

I'd be interested in going through that PR of mine over a call... I could try and finish it first but there are a few questions. I think it needs doing though :see_no_evil:

File: src/routes/tsoa/login.ts (hunk 6 above)
Comment (danielgent) [reply src/routes/tsoa/login.ts:473]: 1. Indeed. I think that PR of mine tries to do some of this but as it's not on production... :smile:  It's not done

2. whatever you think. Do we put more logic in the adapters, less logic, do we get rid of them totally and have SQL statements inline? :man_shrugging:

**Pull request merged**